### PR TITLE
`struct Rav1dState`: Split out the fields of `Rav1dContext` that are mutated on the main thread and put them in a `Mutex`

### PR DESCRIPTION
**Pull request by kkysen**
Our previous code where we passed a `&Rav1dContext` to the worker threads and mutated the `&mut Rav1dContext` through `DAV1D_API`s like `fn rav1d_flush` was unsound, as we were mutating `Rav1dContext` while a `&Rav1dContext` exists.  This moves all of those fields mutated on the main thread from the `DAV1D_API`s into `Rav1dState`, and then puts `Rav1dState` into a `Mutex` (`.try_lock()`ed) into `Rav1dContext`.  We only need to `.try_lock()` it a few places in the `DAV1D_API`s; otherwise, we can just pass `c: &Rav1dContext, state: &mut Rav1dState` args to most `fn`s.  Now `Rav1dContext` is always accessed through a `&` except during its construction in `fn rav1d_open`.

This also lets us remove the `Mutex` around `cached_error_props` and the `Atomic` around `frame_flags`, since they are now accessed through a `&mut Rav1dState`.

### CODE REVIEW

**Review by randomPoison — COMMENTED** [review]
Mostly looks good to me, but give @rinon a chance to look this over before merging.

File: src/lib.rs
Comment (kkysen) [reply src/lib.rs:304]: Oh, I think I messed up in #1176 slightly, as it was getting set to null by `close_internal` in some code paths.  Not sure if we need to keep that behavior, as it wasn't actually happening in all of the code paths (only happening after we allocated the `Rav1dContext`, so for the default allocator cookie check, or in C for the allocation failure, too).  Probably a bit safer to just set it to null, though, as there's not much downside in doing so.  What do you think?

File: src/lib.rs
Comment (kkysen) [reply src/lib.rs:631]: No, you're right.  I missed those.  I'd like to remove some more of these restored fields, though, as I think we don't actually need to restore them, but I want to do that in a follow-up PR.

Comment (kkysen) [reply src/lib.rs]: Fixed it.  Also simplified it a bit (we don't need to copy the fields out first one-by-one, we can just re-assigned them from `old_state`).

File: src/lib.rs
Comment (randomPoison) [reply src/lib.rs:304]: Okay makes sense, if the original logic was setting it to null then doing so here seems fine.

**Review by randomPoison — APPROVED** [review]
Updates look good to me, just needs a high-level review from @rinon before merging.

**Review by rinon — APPROVED** [review]

**Pull request merged**
